web/packages/teleterm/src/ui/Search/SearchContext.test.tsx

```diff
@@ -225,3 +212,91 @@ describe('closeWithoutRestoringFocus', () => {
     expect(previouslyActive.focus).not.toHaveBeenCalled();
   });
 });
+
+test('search bar state is adjusted to the active document', () => {
+  const rootClusterUri = '/clusters/localhost';
+  const appContext = new MockAppContext();
+  appContext.workspacesService.setState(draftState => {
+    draftState.rootClusterUri = rootClusterUri;
+    draftState.workspaces[rootClusterUri] = {
+      localClusterUri: rootClusterUri,
+      documents: [],
+      location: undefined,
+      accessRequests: undefined,
+    };
+  });
+  const docService =
+    appContext.workspacesService.getActiveWorkspaceDocumentService();
+  const { result } = renderUseSearchContextHook(appContext);
+
+  // initial state, no document
+  expect(result.current.inputValue).toBe('');
+  expect(result.current.filters).toEqual([]);
+  expect(result.current.advancedSearchEnabled).toBe(false);
+
+  // document changes to the cluster document
+  act(() => {
+    const clusterDoc = docService.createClusterDocument({
+      clusterUri: rootClusterUri,
+      queryParams: {
+        search: 'foo',
+        kinds: ['db'],
+        sort: { dir: 'ASC', fieldName: 'name' },
+        advancedSearchEnabled: true,
+      },
+    });
+    docService.add(clusterDoc);
+    docService.open(clusterDoc.uri);
+  });
+
+  expect(result.current.inputValue).toBe('foo');
+  expect(result.current.filters).toEqual([
+    { filter: 'resource-type', resourceType: 'db' },
+  ]);
+  expect(result.current.advancedSearchEnabled).toBe(true);
+
+  // document changes to another cluster document
+  act(() => {
+    const clusterDoc = docService.createClusterDocument({
+      clusterUri: rootClusterUri,
+      queryParams: {
+        search: 'bar',
+        kinds: ['kube_cluster'],
+        sort: { dir: 'ASC', fieldName: 'name' },
+        advancedSearchEnabled: false,
+      },
+    });
+    docService.add(clusterDoc);
+    docService.open(clusterDoc.uri);
+  });
+
+  expect(result.current.inputValue).toBe('bar');
+  expect(result.current.filters).toEqual([
+    { filter: 'resource-type', resourceType: 'kube_cluster' },
+  ]);
+  expect(result.current.advancedSearchEnabled).toBe(false);
+
+  // document changes to a non-cluster document
+  act(() => {
+    const clusterDoc = docService.createTshNodeDocument(
+      '/clusters/abc/servers/bar',
+      { origin: 'search_bar' }
+    );
+    docService.add(clusterDoc);
+    docService.open(clusterDoc.uri);
+  });
+
+  expect(result.current.inputValue).toBe('');
+  expect(result.current.filters).toEqual([]);
+  expect(result.current.advancedSearchEnabled).toBe(false);
+});
+
+function renderUseSearchContextHook(appContext?: IAppContext) {
+  return renderHook(() => useSearchContext(), {
```

Comment:
Just my opinion and a nit for the future: I think it's best to avoid abstracting away calls to the thing under tests as well as calls to 3rd party testing utilities which are needed to actually write a particular test.

From the perspective of someone new to the codebase, there are two reasons for avoiding this:

1\. We're in `SearchContext.test.tsx` so want to know where in tests `useSearchContext` is called. Grepping the file for `useSearchContext` will reveal that we have to look for `renderUseSearchContextHook` instead.

2\. `renderHook` is a function from a 3rd party package that provides certain utilities like `result`, `waitFor` etc. When I'm reading a test and the call to `renderHook` is abstracted by a custom function, I no longer have the confidence that `result`, `waitFor` etc. do what they usually do as that custom function might actually alter the behavior of them so again, I have to check what `renderUseSearchContextHook` does exactly.

An alternative way of solving the problem of providing a set of wrappers would be to define a `Wrapper` component and pass it to each call of `renderHook`.

In general from my experience the fewer abstractions in tests the easier they're to maintain.

Reply:
These are good arguments. I replaced my abstraction with a `Wrapper` component.